toucan_connectors/toucan_connector.py

```diff
@@ -504,3 +505,68 @@ def format_db_model(
             .reset_index()
             .to_dict('records')
         )
+
+
+class MalformattedVersion(Exception):
+    """raise when the given version of the engine is not well formated"""
+
+
+class UnavailableVersion(Exception):
+    """raise when the version of the engine is not available"""
+
+
+class VersionableEngineConnector(ABC):
+    """
+    This class is responsable for formating and rendering the version engine
+    from which the connector is fetching data from !
+
+    We suppose all versions should fit the semver format and for that, we
+    build a regex match that can be describe in 6 points :
+    - Major version is required, minor and patch version, and the meta version are supported but optional.
+    - Major, minor, and/or patch can be 0 but cannot be a non-zero value with a leading 0.
+    - The meta version can contain any Unicode letter (not restricted to Latin characters), _, ., and - characters.
+    - Minor version values must suffix to a major version.
+    - Patch version values must suffix to a minor version.
+    - Meta version values can suffix to a minor or patch version.
+
+    Match Groups:
+
+    1- Major
+    2- Minor (optional - requires major to match)
+    3- Patch (optional - requires minor to match)
+    4- Meta/pre-release (optional - requires minor to match)
+    You can add the ^ prefix and $suffix to do match enforcement.
+    """
+
+    # The output of these rules :
+    semver_regex = (
+        r'(0|(?:[1-9]\d*))(?:\.(0|(?:[1-9]\d*))(?:\.(0|(?:[1-9]\d*)))?(?:\-([\w][\w\.\-_]*))?)?'
+    )
+
+    def _validate(self, engine_version: str | float | None) -> bool:
+        """
+        A small validation function for incoming version format
+        """
+        if engine_version is None:
+            raise UnavailableVersion  # pragma: no cover
+
+        return re.match(self.semver_regex, str(engine_version)) is not None
+
+    @abstractmethod
+    def get_engine_version(self) -> tuple:
+        """This method will return the version of the target engine"""
+
+    def _format_version(self, input_version: str | float | None = None) -> tuple:
```

Comment:
Could you please add a unit test with a lot of different inputs for this ? Like `1`, `'1'`, `1.2`, `'1.2'`, `'14.3 (Debian 14.3-1.pgdg110+1)'` etc ?

Reply:
[`62b6294` (#662)](https://github.com/ToucanToco/toucan-connectors/pull/662/commits/62b629407cf559a8e539308321e3d7a513f68afb)